docs/notebooks/How_JAX_primitives_work.ipynb

```diff
@@ -0,0 +1,1651 @@
+{
+  "nbformat": 4,
+  "nbformat_minor": 0,
+  "metadata": {
+    "colab": {
+      "name": "How JAX primitives work.ipynb",
+      "provenance": [],
+      "collapsed_sections": [],
+      "toc_visible": true
+    },
+    "kernelspec": {
+      "name": "python3",
+      "display_name": "Python 3"
+    }
+  },
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "vfxqky4PCUnh",
+        "colab_type": "text"
+      },
+      "source": [
+        "# How JAX primitives work\n",
+        "\n",
+        "*necula@google.com*, October 2019.\n",
+        "\n",
+        "JAX implements certain transformations of Python functions, e.g., `jit`, `grad`,\n",
+        "`vmap`, or `pmap`. The Python functions to be transformed must be JAX-traceable, \n",
+        "which means that as the Python function executes\n",
+        "the only operations it applies to the data are either inspections of data\n",
+        "attributes such as shape or type, or special operations called JAX primitives.\n",
+        "In particular, a JAX-traceable function is sometimes invoked by JAX with\n",
+        "abstract arguments. An example of a JAX abstract value is `ShapedArray(float32[2,2])`, \n",
+        "which captures the type and the shape of values, but not the concrete data values.\n",
+        "JAX primitives know how to operate on both concrete data\n",
+        "values and on the JAX abstract values.\n",
+        "\n",
+        "\n",
+        "The JAX-transformed functions must themselves be JAX-traceable functions,\n",
+        "to ensure that these transformations\n",
+        "can be composed, e.g., `jit(jacfwd(grad(f)))`.\n",
+        "\n",
+        "There are pre-defined JAX primitives corresponding to most XLA operations, \n",
+        "e.g., add, matmul, sin, cos, indexing.\n",
+        "JAX comes with an implementation of numpy functions in terms of JAX primitives, which means that Python programs\n",
+        "using JAX’s implementation of numpy are JAX-traceable and therefore transformable.\n",
+        "Other libraries can be made JAX-traceable by implementing them in terms of JAX primitives.\n",
+        "\n",
+        "The set of JAX primitives is extensible. Instead of reimplementing a function in terms of pre-defined JAX primitives,\n",
+        "one can define a new primitive that encapsulates the behavior of the function.\n",
+        "\n",
+        "**The goal of this document is to explain the interface that a JAX primitive must support in order to allow JAX to perform all its transformations.**\n",
```

Comment:
I'd add a note to the effect that this interface isn't expected to be stable, and might change with revisions to JAX's core.

Reply:
I would rather not add this note. This colab is executable, so if the interface changes either the code will break (and we will know), or the trace will change.